src/java.base/share/classes/java/lang/foreign/MemoryLayout.java

```diff
@@ -594,6 +629,36 @@ static PathElement sequenceElement() {
             return new LayoutPath.PathElementImpl(PathKind.SEQUENCE_ELEMENT,
                                                   LayoutPath::sequenceElement);
         }
+
+        /**
+         * Returns a path element which dereferences an address layout as its
+         * {@linkplain OfAddress#targetLayout() target layout} (where set).
+         * The path element returned by this method does not alter the number of free dimensions of any path
+         * that is combined with such element. Using this path layout to dereference an address layout
+         * that has no target layout results in an {@link IllegalArgumentException} (e.g. when
+         * a var handle is {@linkplain #varHandle(PathElement...) obtained}).
+         *
+         * @return a path element which dereferences an address layout.
+         */
+        static PathElement dereferenceElement() {
+            return new LayoutPath.PathElementImpl(PathKind.DEREF_ELEMENT,
+                    LayoutPath::derefElement);
+        }
+
+        /**
+         * Returns a path element which dereferences an address layout as the given layout.
+         * The path element returned by this method does not alter the number of free dimensions of any path
+         * that is combined with such element. Using this path layout to dereference an address layout
+         * that already has a target layout results in an {@link IllegalArgumentException} (e.g. when
+         * a var handle is {@linkplain #varHandle(PathElement...) obtained}).
+         *
+         * @param layout the
+         * @return a path element which dereferences an address layout.
+         */
+        static PathElement dereferenceElement(MemoryLayout layout) {
```

Comment:
This method needs to be restricted, I believe. Since it performs a resizing of a MemorySegment. The other overload of `dereferenceElement` is already covered because attaching a pointee layout to an address layout is restricted already.

Consider the following:

```
StructLayout hasAddr = MemoryLayout.structLayout(ADDRESS.withName("addr"));

MemorySegment rawSeg = MemorySegment.allocateNative(0, SegmentScope.auto()); // inaccessible segment
// rawSeg.get(JAVA_INT, 0); // not allowed
MemorySegment addrHolder = MemorySegment.allocateNative(hasAddr, SegmentScope.auto());
addrHolder.set(ADDRESS, 0, rawSeg); // fine
VarHandle derefHandle = hasAddr.varHandle(groupElement("addr"), dereferenceElement(JAVA_INT));
int x = derefHandle.get(addrHolder); // oops!
```

Assuming I eye-balled this code correctly, I think this is problematic?

Reply:
Sure - the layout accepting one needs to be restricted - I started off with only the other one (which doesn't need to be restricted), then added the layout accepting one, and forgot to add the restricted check. Any way - as a validation, do we agree that it is handy to have both forms? If you have a layout like A -> B -> A, there is no way to attach target layouts statically - which is the main reason why I added that variant.

Reply:
Yes, it seems good to have it for that case.

Reply:
In the end, I've decided to drop the overload, at least for now. While it could be useful in some pathological situation, the other method should be good for the vast majority of cases. It occurred to me that there are other possible ways to attack the problem of cyclic references - for instance, we could still require a target layout on the address - but then just perform a "layout cast" when dereferencing (which then got me thinking as to whether a more general "cast element" would be required). Overall, I believe it would be better to think a bit more on this one.